x/exchange/fulfillment_test.go

```diff
@@ -0,0 +1,6220 @@
+		{name: "nil", f: orderFulfillment{Order: NewOrder(888)}, exp: false},
+		{name: "unknown", f: orderFulfillment{Order: newUnknownOrder(9)}, exp: false},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual bool
+			testFunc := func() {
+				actual = tc.f.IsBidOrder()
+			}
+			require.NotPanics(t, testFunc, "IsBidOrder()")
+			assert.Equal(t, tc.exp, actual, "IsBidOrder() result")
+		})
+	}
+}
+
+func TestOrderFulfillment_GetMarketID(t *testing.T) {
+	askOrder := func(marketID uint32) orderFulfillment {
+		return orderFulfillment{
+			Order: NewOrder(999).WithAsk(&AskOrder{MarketId: marketID}),
+		}
+	}
+	bidOrder := func(marketID uint32) orderFulfillment {
+		return orderFulfillment{
+			Order: NewOrder(999).WithBid(&BidOrder{MarketId: marketID}),
+		}
+	}
+
+	tests := []struct {
+		name string
+		f    orderFulfillment
+		exp  uint32
+	}{
+		{name: "ask zero", f: askOrder(0), exp: 0},
+		{name: "ask one", f: askOrder(1), exp: 1},
+		{name: "ask five", f: askOrder(5), exp: 5},
+		{name: "ask max uint16+1", f: askOrder(65_536), exp: 65_536},
+		{name: "ask max uint32", f: askOrder(4_294_967_295), exp: 4_294_967_295},
+		{name: "bid zero", f: bidOrder(0), exp: 0},
+		{name: "bid one", f: bidOrder(1), exp: 1},
+		{name: "bid five", f: bidOrder(5), exp: 5},
+		{name: "bid max uint16+1", f: bidOrder(65_536), exp: 65_536},
+		{name: "bid max uint32", f: bidOrder(4_294_967_295), exp: 4_294_967_295},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual uint32
+			testFunc := func() {
+				actual = tc.f.GetMarketID()
+			}
+			require.NotPanics(t, testFunc, "GetMarketID()")
+			assert.Equal(t, tc.exp, actual, "GetMarketID() result")
+		})
+	}
+}
+
+func TestOrderFulfillment_GetOwner(t *testing.T) {
+	askOrder := func(seller string) orderFulfillment {
+		return orderFulfillment{
+			Order: NewOrder(999).WithAsk(&AskOrder{Seller: seller}),
+		}
+	}
+	bidOrder := func(buyer string) orderFulfillment {
+		return orderFulfillment{
+			Order: NewOrder(999).WithBid(&BidOrder{Buyer: buyer}),
+		}
+	}
+	owner := sdk.AccAddress("owner_______________").String()
+
+	tests := []struct {
+		name string
+		f    orderFulfillment
+		exp  string
+	}{
+		{name: "ask empty", f: askOrder(""), exp: ""},
+		{name: "ask not a bech32", f: askOrder("owner"), exp: "owner"},
+		{name: "ask beche32", f: askOrder(owner), exp: owner},
+		{name: "bid empty", f: bidOrder(""), exp: ""},
+		{name: "bid not a bech32", f: bidOrder("owner"), exp: "owner"},
+		{name: "bid beche32", f: bidOrder(owner), exp: owner},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			var actual string
+			testFunc := func() {
+				actual = tc.f.GetOwner()
+			}
+			require.NotPanics(t, testFunc, "GetOwner()")
+			assert.Equal(t, tc.exp, actual, "GetOwner() result")
+		})
+	}
+}
+
+func TestOrderFulfillment_GetAssets(t *testing.T) {
+	coin := func(amt int64) sdk.Coin {
```

Comment:
Same as above for all of these. You could even consider a factory for all your different types of anons.

Reply:
A factory is significant overkill for most of this stuff and would then also require extra testing to make sure the factory is making what you're expecting it to make.